### PR TITLE
[@awarns/ml-kit] Add support for running ml models.

### PR DESCRIPTION
**Pull request by matey97**
This PR adds a way to run machine learning models for regression and classification tasks. The machine learning models supported are TensorFlow Lite models with a CNN (*Convolutional Neural Network*) or a MLP (*Multilayer Perceptron*) architecture.

### CODE REVIEW

File: packages/ml-kit/README.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  // stressLevelRegression
+]
+```
+
+**Task generator parameters:**
+
+| Parameter name                                   | Type                                                  | Description                                                                                                                                                         |
+|--------------------------------------------------|-------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| <code>{classification&vert;regression}Aim</code> | `string`                                              | Objective of the classification/regression. Used to name the task.                                                                                                  |
+| `modelName`                                      | <code>string &vert; ModelNameResolver</code>          | Name of the model (without *tflite* extension) stored in the *ml-models* folder to use for this task, or a function that returns the name of the model when called. |
```
Comment (agonper) [inline packages/ml-kit/README.md:74]: I would put the text between () in bold

File: packages/ml-kit/README.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+| `tag` *(Optional)*                               | `string`                                              | Adds a tag to the name of the task to differentiate it from other tasks with other configurations.                                                                  |
+| `modelOptions` *(Optional)*                      | <code>ModelOptions &vert; ModelOptionsResolver</code> | Configuration to use with the model or a function that returns the configuration when called.                                                                       |
```
Comment (agonper) [inline packages/ml-kit/README.md:76]: For both resolvers, I would give a hint regarding why one would like to use a function instead a plain value (to change it at runtime)

File: packages/ml-kit/README.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+- `ModelNameResolver: () => string`
+
+Useful to change the model used by a task at runtime. You can use the [`ModelManager`](#modelmanager) to obtain a list with the models that are available in the device.
```
Comment (agonper) [inline packages/ml-kit/README.md:80]: Oh, I see it's here too. I don't know what's the best place to put it, to be honest

File: packages/ml-kit/README.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+- `ModelOptions`
+
+| Property       | Type                                    | Description                                                                                                                                                                                                                                                                                               |
+|----------------|-----------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `acceleration` | <code>DelegateType &vert; number</code> | Which type of acceleration to use when running the model. It can take the values `DelegateType.GPU` (GPU acceleration), `DelegateType.NNAPI` ([Android Neural Networks API](https://developer.android.com/ndk/guides/neuralnetworks) acceleration) or a number indicating the quantity of threads to use. |
+
+- `ModelNameResolver: () => ModelOptions`
```
Comment (agonper) [inline packages/ml-kit/README.md:88]: ModelOptionsResolver?

Comment (agonper) [inline packages/ml-kit/README.md:88]: The bullets look a little bit strange to me. Maybe better to user headings?

File: packages/ml-kit/README.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+> ```
+> 
+> **Note**: To use the `writeRecords` task, the persistence package must be installed and configured. See [persistence package docs](https://github.com/GeoTecINIT/awarns-framework/blob/main/packages/persistence/README.md).
+
+### Events
+
+| Name                           | Payload                             | Description |
+|--------------------------------|-------------------------------------|-------------|
+| `{classificationAim}Predicted` | [`Classification`](#classification) |             |
+| `{regressionAim}Predicted`     | [`Regression`](#regression)         |             |
```
Comment (agonper) [inline packages/ml-kit/README.md:114]: Missing descriptions

File: packages/ml-kit/README.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+#### Classification
+
+| Property               | Type                                            | Description                                                |
+|------------------------|-------------------------------------------------|------------------------------------------------------------|
+| `id`                   | `string`                                        | Record's unique id.                                        |
+| `type`                 | `string`                                        | Always `classification-prediction`.                        |
+| `change`               | `Change`                                        | Always `NONE`.                                             |
+| `timestamp`            | `Date`                                          | The local time when the classification record was created. |
+| `classificationResult` | [`ClassificationResult`](#classificationresult) | Object containing the results of the classification.       |
```
Comment (agonper) [inline packages/ml-kit/README.md:127]: To think if it is possible to add information regarding the aim of the classification, either by replacing the content of the type property or by adding a new property. And also to add some "metadata" information regarding the model that came up with the prediction

File: packages/ml-kit/README.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+### Records
+
+#### Classification
+
+| Property               | Type                                            | Description                                                |
+|------------------------|-------------------------------------------------|------------------------------------------------------------|
+| `id`                   | `string`                                        | Record's unique id.                                        |
+| `type`                 | `string`                                        | Always `classification-prediction`.                        |
+| `change`               | `Change`                                        | Always `NONE`.                                             |
+| `timestamp`            | `Date`                                          | The local time when the classification record was created. |
```
Comment (agonper) [inline packages/ml-kit/README.md:126]: How about: "The local time when the model predicted the classification result?" The "was created" thing seems a bit ambiguous to me.

File: packages/ml-kit/README.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+| `classificationResult` | [`ClassificationResult`](#classificationresult) | Object containing the results of the classification.       |
+
+##### ClassificationResult
+
+| Property     | Type                         | Description                                                                                                                |
+|--------------|------------------------------|----------------------------------------------------------------------------------------------------------------------------|
+| `prediction` | `ClassificationPrediction[]` | Array of the classification predictions generated by the model, containing a `label` and classification `score` per class. |
```
Comment (agonper) [inline packages/ml-kit/README.md:133]: Is the ClassificationPrediction another object? If so, what do you think of adding its API here to (although is there textually in the description)

File: packages/ml-kit/README.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+| `timestamp`  | `Date`                       | The local time when the classification result was computed.                                                                |
```
Comment (agonper) [inline packages/ml-kit/README.md:134]: What's the difference between this timestamp and the top-level record timestamp? Perhaps they're redundant?

File: packages/ml-kit/README.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+#### Regression
+
+| Property           | Type                                    | Description                                            |
+|--------------------|-----------------------------------------|--------------------------------------------------------|
+| `id`               | `string`                                | Record's unique id.                                    |
+| `type`             | `string`                                | Always `regression-prediction`.                        |
+| `change`           | `Change`                                | Always `NONE`.                                         |
+| `timestamp`        | `Date`                                  | The local time when the regression record was created. |
+| `regressionResult` | [`RegressionResult`](#regressionresult) | Object containing the results of the regression.       |
```
Comment (agonper) [inline packages/ml-kit/README.md:145]: Same comments as for the Classification record

File: packages/ml-kit/README.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+##### RegressionResult
+
+| Property     | Type       | Description                                               |
+|--------------|------------|-----------------------------------------------------------|
+| `prediction` | `number[]` | Array of numbers with the results generated by the model. |
+| `timestamp`  | `Date`     | The local time when the regression result was computed.   |
```
Comment (agonper) [inline packages/ml-kit/README.md:152]: Same as above

**Review by agonper — COMMENTED** [review]
Well @matey97, I have no words for this. Pretty amazing job on this. Good that you finally decided to extract this from the TUG app. I think it is going to be really useful for other use cases!

Thank you!

I add some comments below

File: packages/ml-kit/README.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+  regressionTask('stress-level', 'stress_regressor-cnn'),
+  // stressLevelRegression
+]
+```
+
+**Task generator parameters:**
+
+| Parameter name                                   | Type                                                  | Description                                                                                                                                                         |
+|--------------------------------------------------|-------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| <code>{classification&vert;regression}Aim</code> | `string`                                              | Objective of the classification/regression. Used to name the task.                                                                                                  |
```
[the remark this answers is not in the document]
Comment (matey97) [reply packages/ml-kit/README.md:73]: Done! Also, as you suggested, now the aim is being used for the type of the records.

File: packages/ml-kit/README.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+### API
+
+#### ModelManager
+You can obtain the singleton instance of the `ModelManager` calling the `getModelManager()` function.
+
+| Method                                                                           | Return type        | Description                                                    |
+|----------------------------------------------------------------------------------|--------------------|----------------------------------------------------------------|
+| `listModels()`                                                                   | `Promise<Model[]>` | Returns a list of the models that are available for their use. |
+
```
[the remark this answers is not in the document]
Comment (matey97) [reply packages/ml-kit/README.md:35]: I will think a bit more about it and to include information to use the module with the API... Some changes might be required for the sake of clarity...

File: packages/ml-kit/README.md (hunk 3 above)
Comment (matey97) [reply packages/ml-kit/README.md:80]: Yeah, me neither. It could be in both places 🤔

File: packages/ml-kit/README.md (hunk 4 above)
Comment (matey97) [reply packages/ml-kit/README.md:88]: The copy paste hehe 😅

Comment (matey97) [reply packages/ml-kit/README.md:88]: Yeah, I'm not convinced neither, I was following the format of other packages. I've restructured it a bit with sections...

File: packages/ml-kit/README.md (hunk 6 above)
Comment (matey97) [reply packages/ml-kit/README.md:127]: Done! I've add that in the `classificationResult` object, not in the record. Then, if the module is used via API (without records) the info is not lost.

File: packages/ml-kit/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (agonper) [reply packages/ml-kit/README.md:74]: Good!

**Comment by matey97** [comment]
> Well @matey97, I have no words for this. Pretty amazing job on this. Good that you finally decided to extract this from the TUG app. I think it is going to be really useful for other use cases!

Thanks for your review and your comments @agonper, I really hope it to be useful for other cases 😄. I've addressed your main concerns, so it's ready for a quick check 😉.

Thanks in advance!

**Comment by agonper** [comment]
Looks good!! I think it is even better now. Excellent work, @matey97!

I'm going to merge this now.

**Pull request merged**